File: DSA Problems/Backtracking/wordBreak.py
def solve(words, ans, target):
    if len(target) == 0:
        print(ans)
        return

    l = len(target)

    for i in range(l):
        left = target[0: i+1]
        

        if words.count(left) > 0:
            right = target[i+1:]
            # print(left, right)
            solve(words, ans + left + " ", right )


tar = ["Iamaboy","Iamagoodandobedientboy"]
dict = [['I', "boy", "am", "a", "aboy", "Iam"],
        ['I', "boy", "am", "good", "aboy", "and", "obedient", "obe", "Iam", "dient", "ando", "a", "agood"]
       ]


for i in range(2):
    print("Case:",i+1)
    print("Target stirng is: ",tar[i])
    solve(dict[i],"", tar[i])
    print("\n")

File: Algo_Problems/Backtracking/Word Break/wordBreak.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution{
    private: 
        string s;
        vector<string> words;
    public: 
        Solution(string _s, vector<string> _words){
            s = _s;
            words = _words;
        }

        bool findWord(string w){
            for(auto x: words) if(w==x) return true;
            return false;
        }

        void util(string ans, string target){
            if(target.size() == 0) {
                cout<<ans<<endl;
                return;
            }

            for(int i=0;i<target.size();i++){
                string left = target.substr(0, i+1);
                if(findWord(left)){
                    string right = target.substr(i+1, target.size());
                    util(ans+left+" ", right);
                }
            }

            
        }

        void solve(){
            util("", s);
        }
        
};

int main(){
    Solution s("Iamaboy", {"I", "boy", "am", "a", "aboy", "Iam"});
    cout<<"Hello World"<<endl;
    s.solve();
}